org.jacoco.core.test/src/org/jacoco/core/internal/instr/InstrSupportTest.java

```diff
@@ -33,6 +36,51 @@ public void setup() {
 		trace = new TraceMethodVisitor(printer);
 	}
 
+	@Test
+	public void needFrames_should_return_false_for_1_1() {
+		assertFalse(InstrSupport.needsFrames(Opcodes.V1_1));
+	}
+
+	@Test
+	public void needFrames_should_return_false_for_1_2() {
+		assertFalse(InstrSupport.needsFrames(Opcodes.V1_2));
+	}
+
+	@Test
+	public void needFrames_should_return_false_for_1_3() {
+		assertFalse(InstrSupport.needsFrames(Opcodes.V1_3));
+	}
+
+	@Test
+	public void needFrames_should_return_false_for_1_4() {
+		assertFalse(InstrSupport.needsFrames(Opcodes.V1_4));
+	}
+
+	@Test
+	public void needFrames_should_return_false_for_1_5() {
+		assertFalse(InstrSupport.needsFrames(Opcodes.V1_5));
+	}
+
+	@Test
+	public void needFrames_should_return_true_for_1_6() {
+		assertTrue(InstrSupport.needsFrames(Opcodes.V1_6));
+	}
+
+	@Test
+	public void needFrames_should_return_true_for_1_7() {
+		assertTrue(InstrSupport.needsFrames(Opcodes.V1_7));
+	}
+
+	@Test
+	public void needFrames_should_return_true_for_1_8() {
+		assertTrue(InstrSupport.needsFrames(Opcodes.V1_8));
+	}
+
+	@Test
+	public void needFrames_should_return_true_for_1_9() {
+		assertTrue(InstrSupport.needsFrames(Opcodes.V9));
```

Comment:
Added Java 10 (luckily ASM lets us write versions > 9).

As discussed before: I personally prefer to have separate tests to see separate results for separate cases. But if you prefer I can combinde them.

Reply:
> I personally prefer to have separate tests to see separate results for separate cases.

I totally agree with this in general, however think that here just two cases and not ten - stackmap frames were introduced in 1.6 and required since then:

```java
	@Test
	public void needFrames_should_return_false_for_versions_less_than_1_6() {
		...
	}

	@Test
	public void needFrames_should_return_true_for_versions_greater_than_or_equal_to_1_6() {
		...
	}
```

this makes clear accent on a functional aspect rather than on technical, which IMO should be prefereable in tests for understandability and hence maintainability.

> Added Java 10

Strange - I don't see it.

> (luckily ASM lets us write versions > 9).

As a side joke: yes it should - see question https://gitlab.ow2.org/asm/asm/merge_requests/92#note_22288 and answer https://gitlab.ow2.org/asm/asm/merge_requests/92#note_22291 , however this is/was not always possible until fix of corresponding bug in ASM 6.1 - https://gitlab.ow2.org/asm/asm/issues/317807 😉
And in any case we can always patch it up after creation.